src/handlers.rs

```diff
@@ -22,68 +23,152 @@ impl fmt::Display for HandlerError {
-                        return Err(HandlerError::Other(e.clone().into()));
-                    }
-                };
                 if let Some(config) = &config.$name {
-                    Handler::handle_input(&$handler, ctx, config, event, input).await.map_err(HandlerError::Other)?;
+                    $name::handle_input(ctx, config, event, input).await.map_err(HandlerError::Other)?;
                 } else {
                     return Err(HandlerError::Message(format!(
                         "The feature `{}` is not enabled in this repository.\n\
-                         To enable it add its section in the `triagebot.toml` \
-                         in the root of the repository.",
+                        To enable it add its section in the `triagebot.toml` \
+                        in the root of the repository.",
                         stringify!($name)
                     )));
                 }
             })*
+            Ok(())
+        }
+    }
+}
+
+issue_handlers! {
+    autolabel,
+    major_change,
+    notify_zulip,
+}
 
-            if let Err(e) = notification::handle(ctx, event).await {
-                log::error!("failed to process event {:?} with notification handler: {:?}", event, e);
+macro_rules! command_handlers {
+    ($($name:ident: $enum:ident,)*) => {
+        async fn handle_command(ctx: &Context, event: &Event, config: &Arc<Config>, body: &str) -> Result<(), HandlerError> {
+            if let Event::Issue(e) = event {
+                if !matches!(e.action, IssuesAction::Opened | IssuesAction::Edited) {
+                    // no change in issue's body for these events, so skip
+                    log::debug!("skipping event, issue was {:?}", e.action);
+                    return Ok(());
+                }
             }
 
-            if let Err(e) = rustc_commits::handle(ctx, event).await {
-                log::error!("failed to process event {:?} with rustc_commits handler: {:?}", event, e);
+            // TODO: parse multiple commands and diff them
+            let mut input = Input::new(&body, &ctx.username);
+            let command = input.parse_command();
+
+            if let Some(previous) = event.comment_from() {
+                let mut prev_input = Input::new(&previous, &ctx.username);
+                let prev_command = prev_input.parse_command();
+                if command == prev_command {
+                    log::info!("skipping unmodified command: {:?} -> {:?}", prev_command, command);
+                    return Ok(());
+                } else {
+                    log::debug!("executing modified command: {:?} -> {:?}", prev_command, command);
+                }
             }
 
+            match command {
+                $(
+                Command::$enum(Ok(command)) => {
+                    if let Some(config) = &config.$name {
+                        $name::handle_command(ctx, config, event, command);//.await.map_err(HandlerError::Other)?;
```

Comment:
Hm so without looking at the compiler error in full, I'm not sure, but sometimes it helps to extract the future to be awaited into a temp variable and then await it separately. I can try and take a look if you're unable to get it to work.

Reply:
I found that the `Command` enum is not `Send` because it may contain `Err(parser::error::Error)` which is not `Send`. This seems to prevent me from sharing it across `await` even if I use only a part of it which should be `Send`.

I fixed it by requiring `Send` on `parser::error::Error` but I don't understand why its needed.